File: design-pattern-python/design_pattern_python/bridge/display_impl.py
from abc import ABCMeta, abstractmethod


class DisplayImpl(metaclass=ABCMeta):
    @abstractmethod
    def raw_open(self):
        pass

    @abstractmethod
    def raw_print(self):
        pass

    @abstractmethod
    def raw_close(self):
        pass


class StringDisplayImpl(DisplayImpl):
    def __init__(self, string: str) -> None:
        self.__string = string
        self.__width = len(string)

    def raw_open(self) -> None:
        self.print_line()

    def raw_print(self) -> None:
        print("|" + self.__string + "|")

    def raw_close(self) -> None:
        self.print_line()

    def print_line(self):
        print("+", end="")
        for _ in range(self.__width):
            print("-", end="")
        print("+")

File: design-pattern-go/bridge/display_impl/display_impl.go
package display_impl

import "fmt"

type DisplyImpl interface {
	RawOpen()
	RawPrint()
	RawClose()
}

type StringDisplayImpl struct {
	str   string
	width int
}

func NewStringDisplayImpl(str string) DisplyImpl {
	return &StringDisplayImpl{
		str:   str,
		width: len(str),
	}
}

func (s *StringDisplayImpl) RawOpen() {
	s.printLine()
}

func (s *StringDisplayImpl) RawPrint() {
	output := "|" + s.str + "|"
	fmt.Println(output)
}

func (s *StringDisplayImpl) RawClose() {
	s.printLine()
}

func (s *StringDisplayImpl) printLine() {
	fmt.Print("+")
	for i := 0; i < s.width; i++ {
		fmt.Print("-")
	}
	fmt.Println("+")
}
